esp-hal/src/lcd_cam/cam.rs

```diff
@@ -360,60 +312,139 @@ impl<'d, CH: DmaChannel> Camera<'d, CH> {
+    camera: Camera<'d, CH>,
+    buffer_view: BUF::View,
+}
+
+impl<'d, CH: DmaChannel, BUF: DmaRxBuffer> CameraTransfer<'d, CH, BUF> {
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
+        // This peripheral doesn't really "complete". As long the camera (or anything
+        // pretending to be :D) sends data, it will receive it and pass it to the DMA.
+        // This implementation of is_done is an opinionated one. When the transfer is
+        // started, the CAM_STOP_EN bit is set, which tells the LCD_CAM to stop
+        // itself when the DMA stops emptying its async RX FIFO. This will
+        // typically be because the DMA ran out descriptors but there could be other
+        // reasons as well.
+
+        // In the future, a user of esp_hal may not want this behaviour, which would be
+        // a reasonable ask. At which point is_done and wait would go away, and
+        // the driver will stop pretending that this peripheral has some kind of
+        // finish line.
+
+        // For now, most people probably want this behaviour, so it shall be kept for
+        // the sake of familiarity and similarity with other drivers.
+
+        self.camera
+            .lcd_cam
+            .cam_ctrl1()
+            .read()
+            .cam_start()
+            .bit_is_clear()
+    }
 
-        assert!(len % self.bus_width == 0);
+    /// Stops this transfer on the spot and returns the peripheral and buffer.
+    pub fn stop(mut self) -> (Camera<'d, CH>, BUF) {
+        self.stop_peripherals();
+        let (camera, view) = self.release();
+        (camera, BUF::from_view(view))
+    }
+
+    /// Waits for the transfer to stop and returns the peripheral and buffer.
+    ///
+    /// Note: The camera doesn't really "finish" its transfer, so what you're
+    /// really waiting for here is a DMA Error. You typically just want to
+    /// call [Self::stop] once you have the data you need.
+    pub fn wait(mut self) -> (Result<(), DmaError>, Camera<'d, CH>, BUF) {
+        while !self.is_done() {}
+
+        // Stop the DMA as it doesn't know that the camera has stopped.
+        self.camera.rx_channel.stop_transfer();
+
+        // Note: There is no "done" interrupt to clear.
+
+        let (camera, view) = self.release();
+
+        let result = if camera.rx_channel.has_error() {
+            Err(DmaError::DescriptorError)
+        } else {
+            Ok(())
+        };
 
+        (result, camera, BUF::from_view(view))
+    }
+
+    fn release(self) -> (Camera<'d, CH>, BUF::View) {
         unsafe {
-            self.rx_chain.fill_for_rx(circular, ptr as _, len)?;
-            self.rx_channel
-                .prepare_transfer_without_start(DmaPeripheral::LcdCam, &self.rx_chain)?;
+            let camera = core::ptr::read(&self.camera);
+            let view = core::ptr::read(&self.buffer_view);
+            core::mem::forget(self);
```

Comment:
Use `ManuallyDrop::take`

Reply:
I did try that initially but it means using `unsafe` here *and* in the `Drop`.

Reply:
It's _meant_ for this. `ptr::read` is _not_.

Reply:
Alternatively, I guess you could impl Drop on Camera itself and not bother with unsafe code here at all.

Reply:
> It's _meant_ for this. `ptr::read` is _not_.

Fair, I'll update